include/nbl/asset/utils/IShaderCompiler.h

```diff
@@ -0,0 +1,253 @@
+			CFileSystemIncludeLoader(core::smart_refctd_ptr<system::ISystem>&& system);
+
+			std::string getInclude(const system::path& searchPath, const std::string& includeName) const override;
+
+		protected:
+			core::smart_refctd_ptr<system::ISystem> m_system;
+		};
+
+		class NBL_API2 CIncludeFinder : public core::IReferenceCounted
+		{
+		public:
+			CIncludeFinder(core::smart_refctd_ptr<system::ISystem>&& system);
+
+			// ! includes within <>
+			// @param requestingSourceDir: the directory where the incude was requested
+			// @param includeName: the string within <> of the include preprocessing directive
+			std::string getIncludeStandard(const system::path& requestingSourceDir, const std::string& includeName) const;
+
+			// ! includes within ""
+			// @param requestingSourceDir: the directory where the incude was requested
+			// @param includeName: the string within "" of the include preprocessing directive
+			std::string getIncludeRelative(const system::path& requestingSourceDir, const std::string& includeName) const;
+
+			inline core::smart_refctd_ptr<CFileSystemIncludeLoader> getDefaultFileSystemLoader() const { return m_defaultFileSystemLoader; }
+
+			void addSearchPath(const std::string& searchPath, const core::smart_refctd_ptr<IIncludeLoader>& loader);
+
+			void addGenerator(const core::smart_refctd_ptr<IIncludeGenerator>& generator);
+
+		protected:
+
+			std::string trySearchPaths(const std::string& includeName) const;
+
+			std::string tryIncludeGenerators(const std::string& includeName) const;
+
+			struct LoaderSearchPath
+			{
+				core::smart_refctd_ptr<IIncludeLoader> loader = nullptr;
+				std::string searchPath = {};
+			};
+
+			std::vector<LoaderSearchPath> m_loaders;
+			std::vector<core::smart_refctd_ptr<IIncludeGenerator>> m_generators;
+			core::smart_refctd_ptr<CFileSystemIncludeLoader> m_defaultFileSystemLoader;
+		};
+
+		enum class E_SPIRV_VERSION : uint32_t
+		{
+			ESV_1_0 = 0x010000u,
+			ESV_1_1 = 0x010100u,
+			ESV_1_2 = 0x010200u,
+			ESV_1_3 = 0x010300u,
+			ESV_1_4 = 0x010400u,
+			ESV_1_5 = 0x010500u,
+			ESV_1_6 = 0x010600u,
+		};
+
+		IShaderCompiler(core::smart_refctd_ptr<system::ISystem>&& system);
+
+		/*
+			@stage shaderStage
+			@targetSpirvVersion spirv version
+			@entryPoint entryPoint
+			@sourceIdentifier String that will be printed along with possible errors as source identifier, and used as include's requestingSrc
+			@outAssembly Optional parameter; if not nullptr, SPIR-V assembly is saved in there.
+			@spirvOptimizer Optional parameter;
+			@logger Optional parameter; used for logging errors/info
+			@includeFinder Optional parameter; if not nullptr, it will resolve the includes in the code
+			@maxSelfInclusionCount used only when includeFinder is not nullptr
+			@genDebugInfo Requests compiler to generate debug info (most importantly objects' names).
+				The engine, while running on OpenGL, won't be able to set push constants for shaders loaded as SPIR-V without debug info.
```

Comment:
anything non-vulkan, basically you can't recover the names of original variables with `CShaderIntrospector` without debug info

By variables we mean names of PC/SSBO/UBO blocks, as they're essentially instantiations of structs with custom packing.